test/test_info.cpp

```diff
@@ -0,0 +1,356 @@
+};
+)";
+	auto ast = tooling::buildASTFromCodeWithArgs(code, compiler_args);
+	auto& context = ast->getASTContext();
+
+	auto const record =
+	  selectFirst<clang::CXXRecordDecl>("record", match(cxxRecordDecl().bind("record"), context));
+	REQUIRE(record != nullptr);
+	REQUIRE(record->method_begin() != record->method_end());
+
+	auto const decl = *record->method_begin();
+	constexpr auto description = "Returns a value."sv;
+	auto parameters = std::vector<info::parameter_info>{};
+	auto const returns = info::return_info{"a value"};
+	auto const preconditions = std::vector<info::precondition_info>{
+	  {"``last`` is reachable from ``first``."},
+	};
+	auto const postconditions = std::vector<info::postcondition_info>{
+	  {"``insert(first, last)`` returns an iterator in the closed interval $[first, last]$."},
+	};
+
+	auto const info = info::function_info(
+	  decl,
+	  std::string(description),
+	  parameters,
+	  returns,
+	  preconditions,
+	  postconditions,
+	  {},
+	  {});
+	CHECK(info.description() == description);
+	CHECK(info.template_parameters().empty());
+	CHECK(info.parameters().empty());
+	CHECK(info.returns() == returns);
+	CHECK(info.exception_specifier().data.empty());
+	CHECK(std::ranges::equal(info.preconditions(), preconditions));
+	CHECK(std::ranges::equal(info.postconditions(), postconditions));
+	CHECK(info.headers().empty());
+	CHECK(info.modules().empty());
+}
+
+TEST_CASE("function templates")
+{
+	constexpr auto code = R"(
+template<class T>
+concept regular = true;
+
+template<class T>
+inline constexpr bool is_nothrow_copyable = true;
+
+template<auto V, class T, regular U, template<class> class W, class... Args>
+auto f(Args&&... args) noexcept((is_nothrow_copyable<Args> and ...));
+)";
+	auto ast = tooling::buildASTFromCodeWithArgs(code, compiler_args);
+	auto& context = ast->getASTContext();
+	auto const decl = selectFirst<clang::FunctionTemplateDecl>(
+	  "decl",
+	  match(functionTemplateDecl().bind("decl"), context));
+	REQUIRE(decl != nullptr);
+	REQUIRE(decl->getAsFunction()->param_size() == 1);
+
+	auto const template_params = decl->getTemplateParameters();
+	REQUIRE(template_params->size() == 5);
+
+	auto const description = "Foosha Village"sv;
+	auto template_parameters = std::vector{
+	  info::template_parameter_info(
+	    llvm::dyn_cast<clang::NonTypeTemplateParmDecl>(template_params->getParam(0)),
+	    "Goat Island"),
+	  info::template_parameter_info(
+	    llvm::dyn_cast<clang::TemplateTypeParmDecl>(template_params->getParam(1)),
+	    "Shells Town"),
+	  info::template_parameter_info(
+	    llvm::dyn_cast<clang::TemplateTypeParmDecl>(template_params->getParam(2)),
+	    "Shimotsuki Village"),
+	  info::template_parameter_info(
+	    llvm::dyn_cast<clang::TemplateTemplateParmDecl>(template_params->getParam(3)),
+	    "Orange Town"),
+	  info::template_parameter_info(
+	    llvm::dyn_cast<clang::TemplateTypeParmDecl>(template_params->getParam(4)),
+	    "Syrup Village"),
+	};
```

Comment:
```suggestion
	auto template_param = []<typename T>(std::size_t i, std::string_view desc, std::type_identity<T>){
		return info::template_parameter_info(
			llvm::dyn_cast<T>(template_params->getParam(i)),
			desc);
	};
	auto template_parameters = std::vector{
		template_param(0, "Goat Island",        std::type_identity<clang::NonTypeTemplateParmDecl>{}),
		template_param(1, "Shells Town",        std::type_identity<clang::TemplateTypeParmDecl>{}),
		template_param(2, "Shimotsuki Village", std::type_identity<clang::TemplateTypeParmDecl>{}),
		template_param(3, "Orange Town",        std::type_identity<clang::TemplateTemplateParmDecl>{}),
		template_param(4, "Syrup Village",      std::type_identity<clang::TemplateTypeParmDecl>{}),
	};
```
Readability wasn't bad before, I thought this may be nicer.

Reply:
Thanks, I like it! Going to change just a little though, since I don't really like `type_identity`.

Reply:
Huh, TIL you can't explicitly pass template parameters when calling a lambda, which makes sense.